pkg/ibmvpc/vpc.go

```diff
@@ -866,40 +867,89 @@ func actionNameStr(action *string) (string, error) {
-	if transitEnablesConn {
-		return noVerboseStr + "allows connection" + newline, nil
-	}
-	return noVerboseStr + "denies connection" + newline, nil
-}
-
-func (tgw *TransitGateway) prefixOfSrcDst(src, dst vpcmodel.Node) *tgwPrefixFilter {
+// RulesInConnectivity returns the prefix filters relevant for <src, dst>. Since src/dst could be a cidr,
+// there could be more than one relevant prefix filter (in a single transit connection)
+// todo: currently "more than one" is not possible since src/dst must be within a given subnet.
+// todo This will be relevant only once we allow src/dst to span over vpcs
+func (tgw *TransitGateway) RulesInConnectivity(src, dst vpcmodel.Node) []vpcmodel.RulesInTable {
 	// <src, dst> routed by tgw given that source is in the tgw,
 	// and there is a prefix filter defined for the dst,
-	// the relevant prefix filter is determined by match of the Address prefix the dest's node is in (including default)
+	// the relevant prefix filters are determined by match of the Address prefix the dest's node is in (including default)
+
+	// vpcsAPToPrefixRules[dst.VPC().UID()] maps from disjoint IPs to RulesInTable
+	// In RulesInTable there is an entry per transit connection; since there could be more than one disjoint IP mapped
+	// to the same transit connection we need to merge the maps from disjoint IPs per transit connection
+	// 1. Collect RulesInTable per transit connection
+	trCnToRulesInTable := map[int][]vpcmodel.RulesInTable{}
 	if vpcmodel.HasNode(tgw.sourceNodes, src) {
-		for _, singleIPBlockPrefix := range tgw.vpcsAPToFilters[dst.VPC().UID()] {
-			if dst.IPBlock().ContainedIn(singleIPBlockPrefix.IPBlock) {
-				return &singleIPBlockPrefix.prefixFilter
+		for ipBlock, trCnPrefixes := range tgw.vpcsAPToPrefixRules[dst.VPC().UID()] {
+			if !dst.IPBlock().Intersect(ipBlock).IsEmpty() {
+				if _, ok := trCnToRulesInTable[trCnPrefixes.Table]; !ok {
+					trCnToRulesInTable[trCnPrefixes.Table] = []vpcmodel.RulesInTable{}
+				}
+				trCnToRulesInTable[trCnPrefixes.Table] = append(trCnToRulesInTable[trCnPrefixes.Table],
+					trCnPrefixes)
 			}
 		}
 	}
-	return nil
+	aggregatedRes := []vpcmodel.RulesInTable{}
+	// 2. Aggregates RulesInTable per transit connection
+	for trCn, trCnRulesInTable := range trCnToRulesInTable {
+		prefixesTrCn := vpcmodel.RulesInTable{Table: trCn, Rules: []int{}, RulesOfType: vpcmodel.NoRules}
+		for _, prefixEntry := range trCnRulesInTable {
+			prefixesTrCn.Rules = append(prefixesTrCn.Rules, prefixEntry.Rules...)
+			if prefixesTrCn.RulesOfType == vpcmodel.NoRules { // first time
+				prefixesTrCn.RulesOfType = prefixEntry.RulesOfType
+			} else if prefixesTrCn.RulesOfType != vpcmodel.BothAllowDeny &&
+				prefixesTrCn.RulesOfType != prefixEntry.RulesOfType {
+				// one is OnlyDeny and the other OnlyAllow
+				prefixesTrCn.RulesOfType = vpcmodel.BothAllowDeny
+			}
+			aggregatedRes = append(aggregatedRes, prefixesTrCn)
```

Comment:
the calculation of is confusing for me.
do we assume that prefixEntry.RulesOfType can not be vpcmodel.BothAllowDeny?

Reply:
I've added explanation. Please see if it is clear now.